### PR TITLE
Accumulate tokens into batches in `PreTrainedTokenizerBase.add_tokens()`

### LINKED ISSUE #16936: Adding tokens to `RobertaTokenizer` is fast, but loading the extended tokenizer from disk takes tens of minutes
### System Info

```shell
- `transformers` version: 4.18.0
- Platform: Linux-5.10.0-0.bpo.9-amd64-x86_64-with-debian-10.12
- Python version: 3.7.3
- Huggingface_hub version: 0.5.1
- PyTorch version (GPU?): 1.11.0+cu102 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: false
- Using distributed or parallel set-up in script?: false
```

### Who can help?

@SaulLu @Narsil

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

I train a BPE tokenizer on a domain-specific dataset and save it as [`tokenizer-latex.json`](https://github.com/huggingface/transformers/files/8557562/tokenizer-latex.json.txt).

``` python
>>> from tokenizers import Tokenizer, normalizers, pre_tokenizers
>>> from tokenizers.models import BPE
>>> from tokenizers.trainers import BpeTrainer
>>> 
>>> latex_model = BPE(unk_token='[UNK]')
>>> latex_tokenizer = Tokenizer(latex_model)
>>> latex_tokenizer.pre_tokenizer = pre_tokenizers.WhitespaceSplit()
>>> latex_tokenizer.normalizer = normalizers.Sequence([normalizers.Strip()])
>>> latex_tokenizer_trainer = BpeTrainer(special_tokens=['[UNK]'])
>>> latex_tokenizer.train(['dataset-latex.txt'], latex_tokenizer_trainer)
>>> latex_tokenizer.save('tokenizer-latex.json')
```

Then, I extend [the pre-trained `roberta-base` tokenizer][1] with 28,141 new tokens from the vocabulary of my BPE tokenizer and I save the result to the directory `./extended-roberta-base/`. This finishes in a matter of seconds:

``` python
>>> from tokenizers import Tokenizer
>>> from transformers import RobertaTokenizer
>>> 
>>> latex_tokenizer = Tokenizer.from_file('tokenizer-latex.json')
>>> 
>>> text_latex_tokenizer = RobertaTokenizer.from_pretrained('roberta-base', add_prefix_space=True)
>>> text_latex_tokenizer.add_tokens(list(latex_tokenizer.get_vocab()))
28141
>>> text_latex_tokenizer.save_pretrained('./extended-roberta-base/')
('./extended-roberta-base/tokenizer_config.json', './extended-roberta-base/special_tokens_map.json',
 './extended-roberta-base/vocab.json', './extended-roberta-base/merges.txt',
 './extended-roberta-base/added_tokens.json', './extended-roberta-base/tokenizer.json')
```

However, when I load the extended `roberta-base` tokenizer from the directory `./extended-roberta-base/`, the library constructs a trie (see https://github.com/huggingface/transformers/pull/13220) over the course of ca 20 minutes:

``` python
>>> from transformers import RobertaTokenizer
>>>
>>> text_latex_tokenizer = RobertaTokenizer.from_pretrained('./extended-roberta-base/')
^C
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
    text_latex_tokenizer = RobertaTokenizer.from_pretrained('./extended-roberta-base/')
  File "***/python3.7/site-packages/transformers/tokenization_utils_base.py", line 1787, in from_pretrained
    **kwargs,
  File "***/python3.7/site-packages/transformers/tokenization_utils_base.py", line 1971, in _from_pretrained
    tokenizer.add_tokens(token, special_tokens=bool(token in special_tokens))
  File "***/python3.7/site-packages/transformers/tokenization_utils_base.py", line 945, in add_tokens
    return self._add_tokens(new_tokens, special_tokens=special_tokens)
  File "***/python3.7/site-packages/transformers/tokenization_utils.py", line 444, in _add_tokens
    self._create_trie(self.unique_no_split_tokens)
  File "***/python3.7/site-packages/transformers/tokenization_utils.py", line 454, in _create_trie
    trie.add(token)
  File "***/python3.7/site-packages/transformers/tokenization_utils.py", line 87, in add
    ref = ref[char]
KeyboardInterrupt
```

The time disparity leads me to believe that when `RobertaTokenizer.add_tokens()` is called, a trie is either not created or is created extremely fast, whereas when `RobertaTokenizer.from_pretrained()` is called, a trie is created (slowly). Using `RobertaTokenizerFast` instead of `RobertaTokenizer` produces similar results at a similar timescale.

 [1]: https://huggingface.co/roberta-base

### Expected behavior

Both `add_tokens()` and `from_pretrained()` should take comparable amount of time. Either building the trie is important and cannot be sped up, in which case `add_tokens()` should also take roughly 20 minutes, or building the trie is unimportant or can be sped up, in which case `from_pretrained()` should finish in a matter of seconds.

### PR DESCRIPTION
**Pull request by Witiko**
# What does this PR do?

For tokenizers with added tokens that only consist of a small number of special tokens or any number of special tokens with consecutive token IDs, this pull request reduces the time complexity of creating the trie from quadratic to linear. This can easily reduce the time it takes to load a pretrained tokenizer with many added tokens from tens of minutes to seconds.

Fixes #16936.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@SaulLu @Narsil

### CODE REVIEW

File: src/transformers/tokenization_utils_base.py
Comment (Narsil) [inline src/transformers/tokenization_utils_base.py:1972]: Using inner functions with n`onlocal` variables is super dangerous IMO (tricky to detect and fix kind of bugs), we should really try to avoid them if we can.

Is there a reason why:

```python
special_tokens = []
tokens = []
is_last_special = None
for token, index in added_tok_encoder_sorted:
    is_special = token in special_tokens
    if is_special == is_last_special:
        tokens.append(token)
        tokens = []
     elif is_last_special:
        tokenizer.add_tokens(tokens, special_tokens=True)
     else:
         tokenizer.add_tokens(tokens, special_tokens=True)
```

It should achieve the same thing, no ? 

TBH I think an even better fix should exist (but this is definitely good enough once we can get rid of `nonlocal` and inner function.

**Review by Narsil — COMMENTED** [review]
Thank you for this proposal !

I think the intent of this PR should work, there might be a way to make the code itself simpler.

There could be an even better fix, but this is already very good, and probably less likely to trigger BC.

File: src/transformers/tokenization_utils_base.py
Comment (Witiko) [reply src/transformers/tokenization_utils_base.py:1972]: @Narsil Thank you for your review! A better fix would be if we could just accumulate all special and non-special tokens and then call `tokenizer.add_tokens()` twice at the end. However, since we need to need to maintain the order of tokens in the vocab, that's not straightforward. However, unless there is a malicious input, where special tokens and non-special tokens are interleaved, my approach should give the same speed with smaller diff against current code.

Comment (Witiko) [reply src/transformers/tokenization_utils_base.py:1972]: Your code seems easier to read and shorter. I will incorporate it shortly.

Comment (Witiko) [reply src/transformers/tokenization_utils_base.py]: @Narsil I simplified the code following your suggestion in 0cd16a8d0.

**Comment by Witiko** [comment]
@Narsil Thank you for your review! A better fix would be if we could just accumulate all special and non-special tokens and then call `tokenizer.add_tokens()` twice at the end. However, since we need to need to maintain the order of tokens in the vocab, we need to call `tokenizer.add_tokens()` each time we encounter a token of a different type (or significantly rework the current logic).

For malicious input, where special tokens and non-special tokens are uniformly interleaved, my approach would lead to no speed-up compared to the current code. However, for the typical case, where special and non-special tokens form two contiguous regions, my approach leads to to optimum speed and has only small diff against the current code.

File: src/transformers/tokenization_utils_base.py
Comment (SaulLu) [inline src/transformers/tokenization_utils_base.py:1968]: I think we can expend a little more this comment to clarify why we do this (to avoid rebuilding the slow tokenizer sort too often). :smile:

**Review by SaulLu — APPROVED** [review]
Thank you for this addition! After spending some time studying this PR I don't see how it could change the previous behavior.

Thanks for the addition. It will indeed speed up the loading of slow tokenizers with many added tokens.

I just want to ping @LysandreJik , core-maintainer of `transformers` as this PR touches the core of tokenizers. Do you want to do a PR review too?

File: src/transformers/tokenization_utils_base.py
Comment (Witiko) [reply src/transformers/tokenization_utils_base.py]: Thank you. I extended the explanation in 625fcfcbd.

**Comment by Witiko** [comment]
@LysandreJik: Thank you for your review. In https://github.com/huggingface/transformers/issues/16936#issuecomment-1127621789, @Narsil indicated that they don't think they can do a proper review of this PR due to lack of time and familiarity with the code base.

**Comment by Narsil** [comment]
@Witiko , I think everyone now agrees that this PR is a good step.

I'll let @LysandreJik decide when to merge.

**Comment by Witiko** [comment]
This pull request is now linked from [the witiko/mathberta model in the Model Hub][1].

 [1]: https://huggingface.co/witiko/mathberta#how-to-use

**Pull request merged**
